<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Long-read methylation data analysis with NanoMethViz and Bioconductor • NanoMethVizWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Long-read methylation data analysis with NanoMethViz and Bioconductor">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">NanoMethVizWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/workflow.html">Long-read methylation data analysis with NanoMethViz and Bioconductor</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/shians/differential_methylation_workshop" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Long-read methylation data analysis with
NanoMethViz and Bioconductor</h1>
                        <h4 data-toc-skip class="author">Shian Su<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/shians/differential_methylation_workshop/blob/devel/vignettes/workflow.rmd" class="external-link"><code>vignettes/workflow.rmd</code></a></small>
      <div class="hidden name"><code>workflow.rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">dmrseq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/shians/NanoMethViz" class="external-link">NanoMethViz</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>DNA methylation is a key epigenetic regulator of gene expression in
mammals, it involves the addition of a methyl group to the 5’ position
of cytosine residues in CpG dinucleotides. The methylation state of CpG
sites can be influenced by a variety of factors, including environmental
exposures, aging, and disease; it plays a crucial role in the regulation
of gene expression, genomic imprinting, and X-chromosome inactivation.
The methylation status of CpG sites can be measured using a variety of
methods, including bisulfite sequencing, methylation arrays, and
nanopore sequencing.</p>
<p>Oxford Nanopore Technologies (ONT) provides a commercial platform for
nanopore long-read sequencing, a third-generation sequencing technology
that produces high-throughput long-read sequencing data. By using
direct-DNA sequencing without PCR amplification, ONT sequencing is able
to sequence DNA molecules with all their modifications intact, including
DNA methylation. This is in contrast to previous sequencing technologies
which required bisulfite conversion to detect DNA methylation, which has
more complex chemical treatments that may introduce bias.</p>
<p>In this workflow, we will demonstrate how to use the
<code>NanoMethViz</code> package to analyze methylation data generated
by ONT long-read sequencing. The <code>NanoMethViz</code> package
provides a suite of tools for analyzing and visualizing methylation data
generated by ONT sequencing, including functions for loading in
methylation data, visualizing methylation patterns, and exporting data
for identifying differentially methylated regions (DMRs) using other
Bioconductor software such as bsseq, DSS, dmrseq and edgeR.</p>
<p>Before we move on, if you do not already have the input data, please
download it by running the code below. This will download the modbam
files that will serve as the raw data for this workshop. The data will
be stored in the <code>input</code> directory.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"input"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">600</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/records/12747551/files/input.tar.gz?download=1"</span>, <span class="st">"input.tar.gz"</span><span class="op">)</span></span>
<span>    <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span><span class="st">"input.tar.gz"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="st">"input.tar.gz"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="experimental-setup">Experimental Setup<a class="anchor" aria-label="anchor" href="#experimental-setup"></a>
</h2>
<p>The data we will be working with comes from the neural stem cells of
triplicate female mice. The mice were F1 crosses between an Xist KO
129S1 female and CAST male, the resulting offspring have sufficiently
different parental chromosomes to allow effective genome-wide
haplotyping to distinguish the parent-of-origin for long-reads, allowing
for the identification of genes that are imprinted or strain-specific.
The Xist KO of the maternal genome also guarantees that X-inactivation
takes place on the paternal X chromosome, allowing for the study of
X-inactivation. Using Oxford Nanopore Technology’s long-read direct DNA
sequencing platform allows for DNA modification information to be
captured and here we use it to study 5mC methylation in the mouse
genome. For this workshop we will only use a subset of the data focusing
on chromosome 7.</p>
<p><img src="MouseDesign.png" width="600"></p>
</div>
<div class="section level2">
<h2 id="loading-in-data">Loading in data<a class="anchor" aria-label="anchor" href="#loading-in-data"></a>
</h2>
<p>modBAM files are BAM files containing reads with additional tags on
each read storing information about methylation probabilities at sites
of interest, this is the currently preferred output of Dorado, the
primary methylation caller offered by ONT. To use this data in
<code>NanoMethViz</code> we must construct an object of the
<code>ModBamResults</code> class. The <code>ModBamResults</code> class
is a container for storing methylation data from multiple samples, along
with sample annotation and exon annotation. To create this object we
need to provide paths to the modBAM files, a sample table, and
optionally exon annotation.</p>
<p><img src="ModBamResult.png" width="600"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># List the modBAM files by searching for files with the .bam extension in the input directory</span></span>
<span><span class="va">bam_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="st">"input"</span>, pattern <span class="op">=</span> <span class="st">"*bam$"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">bam_files</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "input/nsc_1_mat.bam" "input/nsc_1_pat.bam" "input/nsc_2_mat.bam"</span></span>
<span><span class="co">## [4] "input/nsc_2_pat.bam" "input/nsc_3_mat.bam" "input/nsc_3_pat.bam"</span></span></code></pre>
<p>The samples need to be annotated with information for
<code>NanoMethViz</code> to use in aggregating data within experimental
groups. The sample table must contain <code>sample</code> and
<code>group</code> columns with optional additional columns for further
annotation. The rows of this table must match the order of the input
files. The <code>group</code> column is generally the default grouping
column used in the <code>NanoMethViz</code> package. In this example, we
have grouped our data by haplotype.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"input/sample_anno.tsv"</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">samples</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">##   sample    group   tissue haplotype</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> nsc_1_mat nsc_mat nsc    mat      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> nsc_1_pat nsc_pat nsc    pat      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> nsc_2_mat nsc_mat nsc    mat      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> nsc_2_pat nsc_pat nsc    pat      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> nsc_3_mat nsc_mat nsc    mat      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> nsc_3_pat nsc_pat nsc    pat</span></span></code></pre>
<p>Optionally we can add exon annotation to the object to generate
tracks for gene annotation when plotting genomic regions/genes. The exon
annotation must be a <code>data.frame</code> with columns
<code>gene_id</code>, <code>chr</code>, <code>strand</code>,
<code>start</code>, <code>end</code>, <code>transcript_id</code>, and
<code>symbol</code>. A number of helper functions are provided to
retrieve exon annotations in the correct format for human (hg19, hg38)
and mouse (mm10, GRCm39). In this example, we will use the
<code>get_exons_mm10</code> function to retrieve exon annotations for
the mouse genome (mm10) and filter for chromosome 7.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exon_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/get_exons.html" class="external-link">get_exons_mm10</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">chr</span> <span class="op">==</span> <span class="st">"chr7"</span><span class="op">)</span></span>
<span><span class="va">exon_anno</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 60,146 × 7</span></span></span>
<span><span class="co">##    gene_id   chr   strand    start      end transcript_id symbol </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 100009609 chr7  -      84<span style="text-decoration: underline;">935</span>565 84<span style="text-decoration: underline;">941</span>088         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 100009609 chr7  -      84<span style="text-decoration: underline;">943</span>141 84<span style="text-decoration: underline;">943</span>264         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 100009609 chr7  -      84<span style="text-decoration: underline;">943</span>504 84<span style="text-decoration: underline;">943</span>722         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 100009609 chr7  -      84<span style="text-decoration: underline;">946</span>200 84<span style="text-decoration: underline;">947</span>000         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 100009609 chr7  -      84<span style="text-decoration: underline;">947</span>372 84<span style="text-decoration: underline;">947</span>651         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 100009609 chr7  -      84<span style="text-decoration: underline;">963</span>816 84<span style="text-decoration: underline;">964</span>115         <span style="text-decoration: underline;">60</span>883 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 100009609 chr7  -      84<span style="text-decoration: underline;">935</span>565 84<span style="text-decoration: underline;">941</span>088         <span style="text-decoration: underline;">60</span>884 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 100009609 chr7  -      84<span style="text-decoration: underline;">943</span>141 84<span style="text-decoration: underline;">943</span>264         <span style="text-decoration: underline;">60</span>884 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> 100009609 chr7  -      84<span style="text-decoration: underline;">943</span>504 84<span style="text-decoration: underline;">943</span>722         <span style="text-decoration: underline;">60</span>884 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> 100009609 chr7  -      84<span style="text-decoration: underline;">946</span>200 84<span style="text-decoration: underline;">947</span>000         <span style="text-decoration: underline;">60</span>884 Vmn2r65</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 60,136 more rows</span></span></span></code></pre>
<p>Once we have created the <code>ModBamResult</code> object, we can use
the <code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene()</a></code> function to visualize the methylation
patterns of a gene of interest. In this example, we will plot the
methylation patterns of the <em>Peg3</em> gene on chromosome 7 which is
known to be imprinted in mice. The plot contains a smoothed trend line
of the group-aggregated methylation levels across the gene body, along
with a heatmap showing the methylation levels of individual CpG sites in
each sample, and a track showing the isoforms of the gene if exon
annotation is provided.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mbr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/ModBamResult-class.html" class="external-link">ModBamResult</a></span><span class="op">(</span></span>
<span>    methy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/ModBamFiles.html" class="external-link">ModBamFiles</a></span><span class="op">(</span></span>
<span>        paths <span class="op">=</span> <span class="va">bam_files</span>,</span>
<span>        samples <span class="op">=</span> <span class="va">samples</span><span class="op">$</span><span class="va">sample</span></span>
<span>    <span class="op">)</span>,</span>
<span>    samples <span class="op">=</span> <span class="va">samples</span>,</span>
<span>    exons <span class="op">=</span> <span class="va">exon_anno</span>,</span>
<span>    mod_code <span class="op">=</span> <span class="st">"m"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene</a></span><span class="op">(</span><span class="va">mbr</span>, <span class="st">"Peg3"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_files/figure-html/unnamed-chunk-9-1.png" width="600"></p>
<p>The <em>Peg3</em> (paternally expressed gene 3) gene is known to be
imprinted in mice, with the paternal allele being expressed and the
maternal allele being silenced. In the plot, we see that there is a
strong demthylation pattern on the paternal allele near the
transcription start site (TSS) of the gene, allowing the gene to be
expressed. The maternal allele is hypermethylated near the TSS,
silencing the gene. This is a typical methylation pattern for an
imprinted gene, and consistent with paternal expression of
<em>Peg3</em>.</p>
</div>
<div class="section level2">
<h2 id="genome-wide-analysis">Genome-wide analysis<a class="anchor" aria-label="anchor" href="#genome-wide-analysis"></a>
</h2>
<p>While it is possible to use modBAM files to analyse specific regions
of interest, genome-wide analysis requires the data to be in a
tabix-indexed TSV file as it is easier to parse at scale. We can convert
the modBAM files to a tabix-indexed TSV file using the
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/modbam_to_tabix.html" class="external-link">modbam_to_tabix()</a></code> function. This function will create a
tabix-indexed TSV file containing the methylation data from the modBAM
files. This file can then be used as input to the
<code>NanoMethViz</code> functions that require sorted genomic data. In
this example, we will convert the <code>ModBamResult</code> object to a
tabix-indexed TSV file and save it to the <code>data</code> directory.
We will use a pre-generated file in the interest of time, but you can
run the code below to generate the file.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"data"</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"data/methy.tsv.bgz"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/modbam_to_tabix.html" class="external-link">modbam_to_tabix</a></span><span class="op">(</span><span class="va">mbr</span>, <span class="st">"data/methy.tsv.bgz"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Once we have the tabix-indexed TSV file, we can create a
<code>NanoMethResult</code> object using the <code>NanoMethResult</code>
function. The <code>NanoMethResult</code> object contains the same
information as the <code>ModBamResult</code> object but with the
methylation data stored in a tabix-indexed TSV file. To create this
object we need to provide paths to the tabix-indexed TSV file, along
with the accompanying information we generated before.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nmr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/NanoMethResult-class.html" class="external-link">NanoMethResult</a></span><span class="op">(</span></span>
<span>    methy <span class="op">=</span> <span class="st">"data/methy.tsv.bgz"</span>,</span>
<span>    samples <span class="op">=</span> <span class="va">samples</span>,</span>
<span>    exons <span class="op">=</span> <span class="va">exon_anno</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code>NanoMethResult</code> object behaves in the same way as
<code>ModBamResult</code> for any plotting functions. For example we can
use <code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene()</a></code> to visualize the methylation patterns of
the <em>Peg3</em> gene in exactly the same way but replacing the
<code>ModBamResult</code> object with the <code>NanoMethResult</code>
object. We should see the exact same plot as before.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene</a></span><span class="op">(</span><span class="va">nmr</span>, <span class="st">"Peg3"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_files/figure-html/unnamed-chunk-12-1.png" width="600"></p>
<p><code>NanoMethViz</code> is primarily designed for visualising
methylation data, in order to perform differential methylation analysis
we can use other Bioconductor packages that implement sophisticated
statistical procedures, such as <code>dmrseq</code>. To use the
<code>dmrseq</code> function to identify DMRs between the two groups in
our data, the <code>dmrseq</code> function requires a <code>BSseq</code>
object as input, so we convert the <code>NanoMethResult</code> object to
a <code>BSseq</code> object using the <code>methy_to_bsseq</code>
function. Additionally we filter out regions with low coverage which are
generally not informative for differential methylation analysis, here we
remove any sites that are zero in all of the samples of either
experimental group. The <code>BSSeq</code> object stores the methylation
data as two matrices, the methylation matrix and the coverage matrix,
with the option to store additional information such as sample
annotation and genomic regions.</p>
<p>This conversion process usually takes some time, so we will load a
pre-generated <code>BSSeq</code> object, the code used to generate the
object is provided below.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"data/bss.rds"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">bss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/methy_to_bsseq.html" class="external-link">methy_to_bsseq</a></span><span class="op">(</span><span class="va">nmr</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">bss</span>, <span class="st">"data/bss.rds"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">bss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"data/bss.rds"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">pat_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/bsseq/man/getCoverage.html" class="external-link">getCoverage</a></span><span class="op">(</span><span class="va">bss</span><span class="op">)</span><span class="op">)</span>, <span class="st">"pat"</span><span class="op">)</span></span>
<span><span class="va">mat_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/bsseq/man/getCoverage.html" class="external-link">getCoverage</a></span><span class="op">(</span><span class="va">bss</span><span class="op">)</span><span class="op">)</span>, <span class="st">"mat"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">low_cov</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/bsseq/man/getCoverage.html" class="external-link">getCoverage</a></span><span class="op">(</span><span class="va">bss</span><span class="op">)</span><span class="op">[</span>, <span class="va">pat_cols</span><span class="op">]</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span> <span class="op">|</span></span>
<span>           <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/bsseq/man/getCoverage.html" class="external-link">getCoverage</a></span><span class="op">(</span><span class="va">bss</span><span class="op">)</span><span class="op">[</span>, <span class="va">mat_cols</span><span class="op">]</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">low_cov</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">low_cov</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## low_cov</span></span>
<span><span class="co">##     FALSE      TRUE </span></span>
<span><span class="co">## 0.2853227 0.7146773</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bss</span> <span class="op">&lt;-</span> <span class="va">bss</span><span class="op">[</span><span class="op">!</span><span class="va">low_cov</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="mds-plot">MDS Plot<a class="anchor" aria-label="anchor" href="#mds-plot"></a>
</h2>
<p>Using the <code>BSSeq</code> object we are able to create a
multi-dimensional scaling (MDS) plot to visualize the methylation
patterns of the samples, a method commonly used in differential gene
expression analysis. Since <code>BSSeq</code> objects store the
methylation data as two matrices, the methylation matrix and the
coverage matrix, we need to convert the <code>BSSeq</code> object such
that each CpG site from each sample is represented by a single value in
a matrix. This can be done using the
<code>bsseq_to_log_methy_ratio</code> function which converts the
<code>BSSeq</code> object to a matrix of the log of the methylation
ratio with a small count added to prevent division by zero. The
<code>plot_mds</code> function can then be used to create an MDS plot of
the samples using the log methylation ratio matrix as input. The samples
are colored by group, in this case, maternal and paternal.</p>
<p>Because the number of CpG sites is large, we accumulate the
methylation data over genes to both reduce the size of the data but also
produce a more stable measure of methylation. This is done by providing
the <code>regions = exons_to_genes(exons(nmr))</code> argument to
<code>bsseq_to_log_methy_ratio</code> which will aggregate the
methylation data over genes. Without the argument, the function will use
individual CpG sites. After the data is aggregated,
<code>plot_mds</code> uses only the top 500 most variable features to
create the MDS plot.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cgi_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/get_exons.html" class="external-link">get_cgi_mm10</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">chr</span> <span class="op">==</span> <span class="st">"chr7"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cgi_anno</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 1,193 × 14</span></span></span>
<span><span class="co">##      bin chr     start     end gene_id  length cpgNum gcNum perCpg perGc obsExp</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span>   609 chr7  3<span style="text-decoration: underline;">181</span>031 3<span style="text-decoration: underline;">181</span>267 CpG: 22     236     22   147   18.6  62.3   1   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span>   610 chr7  3<span style="text-decoration: underline;">303</span>442 3<span style="text-decoration: underline;">303</span>817 CpG: 28     375     28   249   14.9  66.4   0.69</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span>   610 chr7  3<span style="text-decoration: underline;">366</span>695 3<span style="text-decoration: underline;">366</span>942 CpG: 27     247     27   152   21.9  61.5   1.22</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span>   611 chr7  3<span style="text-decoration: underline;">411</span>983 3<span style="text-decoration: underline;">412</span>269 CpG: 26     286     26   188   18.2  65.7   0.84</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span>   611 chr7  3<span style="text-decoration: underline;">414</span>874 3<span style="text-decoration: underline;">415</span>855 CpG: 87     981     87   518   17.7  52.8   1.28</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span>   611 chr7  3<span style="text-decoration: underline;">423</span>522 3<span style="text-decoration: underline;">423</span>834 CpG: 28     312     28   215   17.9  68.9   0.87</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span>   612 chr7  3<span style="text-decoration: underline;">617</span>298 3<span style="text-decoration: underline;">617</span>534 CpG: 17     236     17   146   14.4  61.9   0.76</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span>   612 chr7  3<span style="text-decoration: underline;">629</span>691 3<span style="text-decoration: underline;">630</span>056 CpG: 32     365     32   195   17.5  53.4   1.24</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span>   612 chr7  3<span style="text-decoration: underline;">645</span>002 3<span style="text-decoration: underline;">645</span>963 CpG: 106    961    106   652   22.1  67.8   0.96</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span>   612 chr7  3<span style="text-decoration: underline;">665</span>238 3<span style="text-decoration: underline;">665</span>517 CpG: 27     279     27   184   19.4  65.9   0.93</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1,183 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: transcript_id &lt;chr&gt;, strand &lt;chr&gt;, symbol &lt;chr&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lmr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/bsseq_to_log_methy_ratio.html" class="external-link">bsseq_to_log_methy_ratio</a></span><span class="op">(</span></span>
<span>    <span class="va">bss</span>,</span>
<span>    regions <span class="op">=</span> <span class="va">cgi_anno</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lmr</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract</a></span><span class="op">(</span><span class="st">"mat|pat"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_mds.html" class="external-link">plot_mds</a></span><span class="op">(</span><span class="va">lmr</span>, groups <span class="op">=</span> <span class="va">groups</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_files/figure-html/unnamed-chunk-14-1.png" width="600"></p>
<p>From this MDS we see that there is not a clear separation between the
maternal and paternal groups, suggesting that there is no strong global
difference in methylation patterns between the two groups, and that we
don’t expect to find many differentially methylated regions between the
two groups. This is mostly expected as imprinted genes are rare in mice
and most genes in the autosomes should behave similarly between the two
parental alleles.</p>
</div>
<div class="section level2">
<h2 id="aggregating-feature-methylation-data">Aggregating feature methylation data<a class="anchor" aria-label="anchor" href="#aggregating-feature-methylation-data"></a>
</h2>
<p>It is often informative to aggregate methylation data over a class of
features to identify broad patterns of methylation across the features.
This can help establish overall differences in methylation patterns
between groups in a class of features such as genes, CpG islands, or
enhancers. The <code>NanoMethViz</code> package provides a functions for
aggregating methylation data across genomic regions. The
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_agg_regions.html" class="external-link">plot_agg_regions()</a></code> function can be used to visualize the
aggregated methylation data across a set of features of interest defined
by a table of coordinates. For example we may want to investigate the
methylation patterns of a set of genes on chromosome 7. We can use the
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/exons_to_genes.html" class="external-link">exons_to_genes()</a></code> helper function to convert exon
annotations already stored in the object to gene annotations and filter
for only chromosome 7 genes. Then, for speed, we can subset 100 regions
and use the <code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_agg_regions.html" class="external-link">plot_agg_regions()</a></code> function to visualize the
mean methylation profile across those genes. With more time we could use
the full set of genes.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/exons_to_genes.html" class="external-link">exons_to_genes</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/exons.html" class="external-link">exons</a></span><span class="op">(</span><span class="va">nmr</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_agg_regions.html" class="external-link">plot_agg_regions</a></span><span class="op">(</span><span class="va">nmr</span>, regions <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_sample</a></span><span class="op">(</span><span class="va">gene_anno</span>, n <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>, group_col <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_files/figure-html/unnamed-chunk-15-1.png" width="600"></p>
<p>In this plot we can see that active genes tend to be demethylated
near the TSS region and hypermethylated throughout the genebody. This is
consistent with the proposed function of methylation in gene regulation,
where methylation in promoter region of genes is associated with gene
silencing, while transcription of active genes is associated with
recruitment of methylating mechanisms to the genebody.</p>
<p>Another class of features of interest are CpG islands, which are
regions of the genome that are rich in CpG dinucleotides. CpG islands
are often associated with gene promoters and are generally unmethylated
in normal cells. We can use the <code><a href="https://rdrr.io/pkg/NanoMethViz/man/get_exons.html" class="external-link">get_cgi_mm10()</a></code> function to
retrieve CpG island annotations for the mouse genome (mm10) and filter
for chromosome 7. We can then use the <code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_agg_regions.html" class="external-link">plot_agg_regions()</a></code>
function to visualize the mean methylation profile across CpG islands on
chromosome 7. We expect to see a clear pattern of demethylation over CpG
islands, which tend to be enriched in the promoter regions of genes.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_agg_regions.html" class="external-link">plot_agg_regions</a></span><span class="op">(</span><span class="va">nmr</span>, <span class="va">cgi_anno</span>, group_col <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_files/figure-html/unnamed-chunk-16-1.png" width="600"></p>
</div>
<div class="section level2">
<h2 id="differential-methylation">Differential methylation<a class="anchor" aria-label="anchor" href="#differential-methylation"></a>
</h2>
<p>Another common analysis is to identify differentially methylated
regions (DMRs) between the experimental groups. It is generally more
informative to identify DMRs rather than individual CpG sites, as DMRs
are more likely to be functionally relevant and biologically
interpretable. However DMRs are not easy to identify, as it requires the
identification of individual significant sites, some determination of
whether nearby sites should be aggregated into a region, and a
statistical test to determine if the region as a whole is differentially
methylated.</p>
<p>The <code>dmrseq</code> package can be used to identify
differentially methylated regions (DMRs) between the two groups in our
data. The <code>dmrseq</code> function from the package requires a
<code>BSseq</code> object as input, along with the name of the covariate
to test for differential methylation. We will test for differential
methylation between the maternal and paternal groups. The
<code>dmrseq</code> function returns a <code>GRanges</code> object
containing the DMRs identified by the function. The <code>GRanges</code>
object contains the genomic coordinates of the DMRs, along with
information about the statistical significance of the DMRs.</p>
<p>There are alternatives to <code>dmrseq</code> for identifying DMRs,
such as <code>bsseq</code>, <code>DSS</code> and <code>edgeR</code>. For
<code>bsseq</code> and <code>DSS</code> the same <code>BSSeq</code>
object can be used as input, while for <code>edgeR</code> the
<code>BSSeq</code> object must be further converted to a
<code>DGEList</code> object using the <code><a href="https://rdrr.io/pkg/NanoMethViz/man/bsseq_to_edger.html" class="external-link">bsseq_to_edger()</a></code> or
<code><a href="https://rdrr.io/pkg/NanoMethViz/man/methy_to_edger.html" class="external-link">methy_to_edger()</a></code> functions. <code>bsseq</code>,
<code>DSS</code> and <code>dmrseq</code> all use a similar approach to
identify DMRs, using a statistical test to compare the methylation
levels between the two groups at each CpG site, followed by a clustering
step to identify regions of the genome with consistent differences in
methylation levels for de novo DMR discovery. While <code>edgeR</code>
requires users to choose the regions to test for differential
methylation, which is reflected by the regions in which they choose to
summarise counts. <code>dmrseq</code> and <code>edgeR</code> are able to
produce p-values for each region while <code>bsseq</code> and
<code>DSS</code> only provide p-values for each CpG site along with an
aggregated area statistic.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pData</span><span class="op">(</span><span class="va">bss</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu">NanoMethViz</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/samples.html" class="external-link">samples</a></span><span class="op">(</span><span class="va">nmr</span><span class="op">)</span><span class="op">$</span><span class="va">group</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"data/regions"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/dmrseq/man/dmrseq.html" class="external-link">dmrseq</a></span><span class="op">(</span><span class="va">bss</span>, testCovariate <span class="op">=</span> <span class="st">"condition"</span>, minNumRegion <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">regions</span>, <span class="st">"data/regions.rds"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"data/regions.rds"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>A number of regions will be produced by <code>dmrseq</code>, this
includes regions were aggregated because they contained significant CpG
sites but failed to reach statistical significance at a region level. In
order to make sense of these results, we generally want to associate the
regions to genes. One way to do that is to use the
<code>plyranges</code> object along with the the gene annotation we
already generated. We want to find DMR regions that overlap within a
certain distance of the transcription start site (TSS) of a gene. We use
<code><a href="https://rdrr.io/pkg/plyranges/man/ranges-anchor.html" class="external-link">anchor_5p()</a></code> and <code>mutate(width = 1)</code> to shrink
the annotation to the TSS site, then we stretch the annotation by 10kb
such that it spans 5kb on either side of the TSS. We then use the
<code><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_intersect()</a></code> function to find the DMRs that are
near a gene’s TSS. This will return a <code>GRanges</code> object
containing the DMRs which overlap with the gene TSS regions, and append
information about the gene to the DMRs.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_anno_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-construct.html" class="external-link">as_granges</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="va">gene_anno</span>, seqnames <span class="op">=</span> <span class="st">"chr"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gene_anno_gr_tss</span> <span class="op">&lt;-</span> <span class="va">gene_anno_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-anchor.html" class="external-link">anchor_5p</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/stretch.html" class="external-link">stretch</a></span><span class="op">(</span><span class="fl">10000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gene_dmr_overlaps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_intersect</a></span><span class="op">(</span><span class="va">regions</span>, <span class="va">gene_anno_gr_tss</span><span class="op">)</span></span>
<span><span class="va">gene_dmr_overlaps</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 868 ranges and 9 metadata columns:</span></span>
<span><span class="co">##         seqnames              ranges strand |         L      area       beta</span></span>
<span><span class="co">##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">##     [1]     chr7     6727308-6731838      * |       157  128.8218   -1.84082</span></span>
<span><span class="co">##     [2]     chr7     6727308-6731838      * |       157  128.8218   -1.84082</span></span>
<span><span class="co">##     [3]     chr7 143294268-143296757      * |       112   92.9051   -1.87215</span></span>
<span><span class="co">##     [4]     chr7   60003279-60005228      * |        60   50.9278   -1.82583</span></span>
<span><span class="co">##     [5]     chr7   60003279-60005228      * |        60   50.9278   -1.82583</span></span>
<span><span class="co">##     ...      ...                 ...    ... .       ...       ...        ...</span></span>
<span><span class="co">##   [864]     chr7 126579062-126580889      * |        27   5.00269  0.0232921</span></span>
<span><span class="co">##   [865]     chr7 126580817-126580889      * |        27   5.00269  0.0232921</span></span>
<span><span class="co">##   [866]     chr7 126579942-126580889      * |        27   5.00269  0.0232921</span></span>
<span><span class="co">##   [867]     chr7   19696925-19697636      * |        23   2.65049 -0.0118438</span></span>
<span><span class="co">##   [868]     chr7   19696925-19697636      * |        23   2.65049 -0.0118438</span></span>
<span><span class="co">##               stat        pval      qval         index     gene_id      symbol</span></span>
<span><span class="co">##          &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;     &lt;IRanges&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##     [1]   -40.6010 5.71657e-05 0.0401684   27111-27267       18616        Peg3</span></span>
<span><span class="co">##     [2]   -40.6010 5.71657e-05 0.0401684   27111-27267       57775       Usp29</span></span>
<span><span class="co">##     [3]   -32.5253 1.14331e-04 0.0401684 910828-910939       63830    Kcnq1ot1</span></span>
<span><span class="co">##     [4]   -20.1018 1.14331e-04 0.0401684 335287-335346       52480      Snhg14</span></span>
<span><span class="co">##     [5]   -20.1018 1.14331e-04 0.0401684 335287-335346       84704       Snurf</span></span>
<span><span class="co">##     ...        ...         ...       ...           ...         ...         ...</span></span>
<span><span class="co">##   [864]  0.0925951    0.996341  0.997876 764165-764191   102465641     Mir7059</span></span>
<span><span class="co">##   [865]  0.0925951    0.996341  0.997876 764165-764191       12752        Cln3</span></span>
<span><span class="co">##   [866]  0.0925951    0.996341  0.997876 764165-764191      171504       Apobr</span></span>
<span><span class="co">##   [867] -0.0795915    0.996456  0.997876   81437-81459       11812       Apoc1</span></span>
<span><span class="co">##   [868] -0.0795915    0.996456  0.997876   81437-81459       11816        Apoe</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>After we identify the DMRs that overlap with the gene TSS regions, we
can filter for the significant DMRs using the <code>qval</code> column.
This is the adjusted p-value for the DMRs, calculated using the
Benjamini-Hochberg procedure to control false-discovery rates for
multiple-testing. We can then select the columns we are interested in,
such as the gene symbol, chromosome, start and end positions of the
DMRs, the strand of the gene, and the q-value of the DMRs.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dmr_regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">gene_dmr_overlaps</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="st">"seqnames"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">signif_regions</span> <span class="op">&lt;-</span> <span class="va">dmr_regions</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">qval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="va">signif_regions</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">symbol</span>, <span class="va">chr</span>, <span class="va">start</span>, <span class="va">end</span>, <span class="va">strand</span>, <span class="va">qval</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 8 × 6</span></span></span>
<span><span class="co">##   symbol   chr       start       end strand   qval</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> Peg3     chr7    6<span style="text-decoration: underline;">727</span>308   6<span style="text-decoration: underline;">731</span>838 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Usp29    chr7    6<span style="text-decoration: underline;">727</span>308   6<span style="text-decoration: underline;">731</span>838 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> Kcnq1ot1 chr7  143<span style="text-decoration: underline;">294</span>268 143<span style="text-decoration: underline;">296</span>757 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> Snhg14   chr7   60<span style="text-decoration: underline;">003</span>279  60<span style="text-decoration: underline;">005</span>228 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> Snurf    chr7   60<span style="text-decoration: underline;">003</span>279  60<span style="text-decoration: underline;">005</span>228 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> Mkrn3    chr7   62<span style="text-decoration: underline;">416</span>953  62<span style="text-decoration: underline;">423</span>032 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">7</span> Cdkn1c   chr7  143<span style="text-decoration: underline;">459</span>739 143<span style="text-decoration: underline;">462</span>038 *      0.040<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">8</span> Peg12    chr7   62<span style="text-decoration: underline;">461</span>469  62<span style="text-decoration: underline;">464</span>053 *      0.040<span style="text-decoration: underline;">2</span></span></span></code></pre>
<p>We see that some DMRs overlap with multiple genes, and here we can
plot the methylations of genes <em>Peg3</em> and <em>Kcnq1ot1</em> which
are known to be imprinted in mice with the significant regions
highlighted using <code><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene()</a></code> with the
<code>anno_regions</code> argument set to the regions we identified.
This will highlight the significant regions in the gene plot in a shaded
band. We can then see in both these genes, as required by our filtering
criteria, the differentially methylated regions fall near the TSS of the
genes.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"NanoMethViz.highlight_col"</span> <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene</a></span><span class="op">(</span><span class="va">nmr</span>, <span class="st">"Peg3"</span>, anno_regions <span class="op">=</span> <span class="va">signif_regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_files/figure-html/unnamed-chunk-20-1.png" width="600"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene</a></span><span class="op">(</span><span class="va">nmr</span>, <span class="st">"Kcnq1ot1"</span>, anno_regions <span class="op">=</span> <span class="va">signif_regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_files/figure-html/unnamed-chunk-20-2.png" width="600"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/NanoMethViz/man/plot_gene.html" class="external-link">plot_gene</a></span><span class="op">(</span><span class="va">nmr</span>, <span class="st">"Cdkn1c"</span>, anno_regions <span class="op">=</span> <span class="va">signif_regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_files/figure-html/unnamed-chunk-20-3.png" width="600"></p>
<p>We have now successfully used dmrseq to statistically identify
differentially methylated regions around known imprinted genes.
Additionally we have used that information to visualise the methylation
data of the genes associated with the DMRs, and we see that it is
primarily near the TSS where differential methylation occurs.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] NanoMethViz_3.1.5           plyranges_1.25.0           </span></span>
<span><span class="co">##  [3] lubridate_1.9.3             forcats_1.0.0              </span></span>
<span><span class="co">##  [5] stringr_1.5.1               dplyr_1.1.4                </span></span>
<span><span class="co">##  [7] purrr_1.0.2                 readr_2.1.5                </span></span>
<span><span class="co">##  [9] tidyr_1.3.1                 tibble_3.2.1               </span></span>
<span><span class="co">## [11] ggplot2_3.5.1               tidyverse_2.0.0            </span></span>
<span><span class="co">## [13] dmrseq_1.25.0               bsseq_1.41.0               </span></span>
<span><span class="co">## [15] SummarizedExperiment_1.35.1 Biobase_2.65.0             </span></span>
<span><span class="co">## [17] MatrixGenerics_1.17.0       matrixStats_1.3.0          </span></span>
<span><span class="co">## [19] GenomicRanges_1.57.1        GenomeInfoDb_1.41.1        </span></span>
<span><span class="co">## [21] IRanges_2.39.1              S4Vectors_0.43.1           </span></span>
<span><span class="co">## [23] BiocGenerics_0.51.0        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] splines_4.4.1                            </span></span>
<span><span class="co">##   [2] BiocIO_1.15.0                            </span></span>
<span><span class="co">##   [3] bitops_1.0-7                             </span></span>
<span><span class="co">##   [4] filelock_1.0.3                           </span></span>
<span><span class="co">##   [5] R.oo_1.26.0                              </span></span>
<span><span class="co">##   [6] XML_3.99-0.17                            </span></span>
<span><span class="co">##   [7] lifecycle_1.0.4                          </span></span>
<span><span class="co">##   [8] vroom_1.6.5                              </span></span>
<span><span class="co">##   [9] lattice_0.22-6                           </span></span>
<span><span class="co">##  [10] magrittr_2.0.3                           </span></span>
<span><span class="co">##  [11] limma_3.61.4                             </span></span>
<span><span class="co">##  [12] sass_0.4.9                               </span></span>
<span><span class="co">##  [13] rmarkdown_2.27                           </span></span>
<span><span class="co">##  [14] jquerylib_0.1.4                          </span></span>
<span><span class="co">##  [15] yaml_2.3.9                               </span></span>
<span><span class="co">##  [16] doRNG_1.8.6                              </span></span>
<span><span class="co">##  [17] DBI_1.2.3                                </span></span>
<span><span class="co">##  [18] RColorBrewer_1.1-3                       </span></span>
<span><span class="co">##  [19] abind_1.4-5                              </span></span>
<span><span class="co">##  [20] zlibbioc_1.51.1                          </span></span>
<span><span class="co">##  [21] R.utils_2.12.3                           </span></span>
<span><span class="co">##  [22] RCurl_1.98-1.16                          </span></span>
<span><span class="co">##  [23] rappdirs_0.3.3                           </span></span>
<span><span class="co">##  [24] cpp11_0.4.7                              </span></span>
<span><span class="co">##  [25] GenomeInfoDbData_1.2.12                  </span></span>
<span><span class="co">##  [26] irlba_2.3.5.1                            </span></span>
<span><span class="co">##  [27] pkgdown_2.1.0                            </span></span>
<span><span class="co">##  [28] permute_0.9-7                            </span></span>
<span><span class="co">##  [29] DelayedMatrixStats_1.27.2                </span></span>
<span><span class="co">##  [30] codetools_0.2-20                         </span></span>
<span><span class="co">##  [31] DelayedArray_0.31.8                      </span></span>
<span><span class="co">##  [32] tidyselect_1.2.1                         </span></span>
<span><span class="co">##  [33] outliers_0.15                            </span></span>
<span><span class="co">##  [34] farver_2.1.2                             </span></span>
<span><span class="co">##  [35] UCSC.utils_1.1.0                         </span></span>
<span><span class="co">##  [36] ScaledMatrix_1.13.0                      </span></span>
<span><span class="co">##  [37] BiocFileCache_2.13.0                     </span></span>
<span><span class="co">##  [38] GenomicAlignments_1.41.0                 </span></span>
<span><span class="co">##  [39] jsonlite_1.8.8                           </span></span>
<span><span class="co">##  [40] annotatr_1.31.0                          </span></span>
<span><span class="co">##  [41] e1071_1.7-14                             </span></span>
<span><span class="co">##  [42] iterators_1.0.14                         </span></span>
<span><span class="co">##  [43] systemfonts_1.1.0                        </span></span>
<span><span class="co">##  [44] foreach_1.5.2                            </span></span>
<span><span class="co">##  [45] dbscan_1.2-0                             </span></span>
<span><span class="co">##  [46] tools_4.4.1                              </span></span>
<span><span class="co">##  [47] ragg_1.3.2                               </span></span>
<span><span class="co">##  [48] Rcpp_1.0.12                              </span></span>
<span><span class="co">##  [49] glue_1.7.0                               </span></span>
<span><span class="co">##  [50] SparseArray_1.5.21                       </span></span>
<span><span class="co">##  [51] mgcv_1.9-1                               </span></span>
<span><span class="co">##  [52] xfun_0.45                                </span></span>
<span><span class="co">##  [53] HDF5Array_1.33.3                         </span></span>
<span><span class="co">##  [54] withr_3.0.0                              </span></span>
<span><span class="co">##  [55] BiocManager_1.30.23                      </span></span>
<span><span class="co">##  [56] fastmap_1.2.0                            </span></span>
<span><span class="co">##  [57] rhdf5filters_1.17.0                      </span></span>
<span><span class="co">##  [58] fansi_1.0.6                              </span></span>
<span><span class="co">##  [59] digest_0.6.36                            </span></span>
<span><span class="co">##  [60] rsvd_1.0.5                               </span></span>
<span><span class="co">##  [61] timechange_0.3.0                         </span></span>
<span><span class="co">##  [62] R6_2.5.1                                 </span></span>
<span><span class="co">##  [63] textshaping_0.4.0                        </span></span>
<span><span class="co">##  [64] colorspace_2.1-0                         </span></span>
<span><span class="co">##  [65] Cairo_1.6-2                              </span></span>
<span><span class="co">##  [66] gtools_3.9.5                             </span></span>
<span><span class="co">##  [67] RSQLite_2.3.7                            </span></span>
<span><span class="co">##  [68] R.methodsS3_1.8.2                        </span></span>
<span><span class="co">##  [69] utf8_1.2.4                               </span></span>
<span><span class="co">##  [70] generics_0.1.3                           </span></span>
<span><span class="co">##  [71] data.table_1.15.4                        </span></span>
<span><span class="co">##  [72] rtracklayer_1.65.0                       </span></span>
<span><span class="co">##  [73] class_7.3-22                             </span></span>
<span><span class="co">##  [74] httr_1.4.7                               </span></span>
<span><span class="co">##  [75] htmlwidgets_1.6.4                        </span></span>
<span><span class="co">##  [76] S4Arrays_1.5.4                           </span></span>
<span><span class="co">##  [77] org.Mm.eg.db_3.19.1                      </span></span>
<span><span class="co">##  [78] regioneR_1.37.0                          </span></span>
<span><span class="co">##  [79] pkgconfig_2.0.3                          </span></span>
<span><span class="co">##  [80] gtable_0.3.5                             </span></span>
<span><span class="co">##  [81] blob_1.2.4                               </span></span>
<span><span class="co">##  [82] XVector_0.45.0                           </span></span>
<span><span class="co">##  [83] htmltools_0.5.8.1                        </span></span>
<span><span class="co">##  [84] scales_1.3.0                             </span></span>
<span><span class="co">##  [85] png_0.1-8                                </span></span>
<span><span class="co">##  [86] knitr_1.48                               </span></span>
<span><span class="co">##  [87] tzdb_0.4.0                               </span></span>
<span><span class="co">##  [88] reshape2_1.4.4                           </span></span>
<span><span class="co">##  [89] rjson_0.2.21                             </span></span>
<span><span class="co">##  [90] nlme_3.1-165                             </span></span>
<span><span class="co">##  [91] curl_5.2.1                               </span></span>
<span><span class="co">##  [92] bumphunter_1.47.0                        </span></span>
<span><span class="co">##  [93] proxy_0.4-27                             </span></span>
<span><span class="co">##  [94] cachem_1.1.0                             </span></span>
<span><span class="co">##  [95] rhdf5_2.49.0                             </span></span>
<span><span class="co">##  [96] BiocVersion_3.20.0                       </span></span>
<span><span class="co">##  [97] parallel_4.4.1                           </span></span>
<span><span class="co">##  [98] vipor_0.4.7                              </span></span>
<span><span class="co">##  [99] AnnotationDbi_1.67.0                     </span></span>
<span><span class="co">## [100] ggrastr_1.0.2                            </span></span>
<span><span class="co">## [101] restfulr_0.0.15                          </span></span>
<span><span class="co">## [102] desc_1.4.3                               </span></span>
<span><span class="co">## [103] pillar_1.9.0                             </span></span>
<span><span class="co">## [104] grid_4.4.1                               </span></span>
<span><span class="co">## [105] vctrs_0.6.5                              </span></span>
<span><span class="co">## [106] BiocSingular_1.21.2                      </span></span>
<span><span class="co">## [107] dbplyr_2.5.0                             </span></span>
<span><span class="co">## [108] beachmat_2.21.4                          </span></span>
<span><span class="co">## [109] beeswarm_0.4.0                           </span></span>
<span><span class="co">## [110] evaluate_0.24.0                          </span></span>
<span><span class="co">## [111] GenomicFeatures_1.57.0                   </span></span>
<span><span class="co">## [112] cli_3.6.3                                </span></span>
<span><span class="co">## [113] locfit_1.5-9.10                          </span></span>
<span><span class="co">## [114] compiler_4.4.1                           </span></span>
<span><span class="co">## [115] Rsamtools_2.21.0                         </span></span>
<span><span class="co">## [116] rlang_1.1.4                              </span></span>
<span><span class="co">## [117] crayon_1.5.3                             </span></span>
<span><span class="co">## [118] rngtools_1.5.2                           </span></span>
<span><span class="co">## [119] labeling_0.4.3                           </span></span>
<span><span class="co">## [120] plyr_1.8.9                               </span></span>
<span><span class="co">## [121] fs_1.6.4                                 </span></span>
<span><span class="co">## [122] ggbeeswarm_0.7.2                         </span></span>
<span><span class="co">## [123] TxDb.Mmusculus.UCSC.mm10.knownGene_3.10.0</span></span>
<span><span class="co">## [124] stringi_1.8.4                            </span></span>
<span><span class="co">## [125] BiocParallel_1.39.0                      </span></span>
<span><span class="co">## [126] assertthat_0.2.1                         </span></span>
<span><span class="co">## [127] munsell_0.5.1                            </span></span>
<span><span class="co">## [128] Biostrings_2.73.1                        </span></span>
<span><span class="co">## [129] Matrix_1.7-0                             </span></span>
<span><span class="co">## [130] BSgenome_1.73.0                          </span></span>
<span><span class="co">## [131] hms_1.1.3                                </span></span>
<span><span class="co">## [132] patchwork_1.2.0                          </span></span>
<span><span class="co">## [133] sparseMatrixStats_1.17.2                 </span></span>
<span><span class="co">## [134] bit64_4.0.5                              </span></span>
<span><span class="co">## [135] Rhdf5lib_1.27.0                          </span></span>
<span><span class="co">## [136] KEGGREST_1.45.1                          </span></span>
<span><span class="co">## [137] statmod_1.5.0                            </span></span>
<span><span class="co">## [138] highr_0.11                               </span></span>
<span><span class="co">## [139] AnnotationHub_3.13.0                     </span></span>
<span><span class="co">## [140] memoise_2.0.1                            </span></span>
<span><span class="co">## [141] bslib_0.7.0                              </span></span>
<span><span class="co">## [142] bit_4.0.5</span></span></code></pre>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:su.s@wehi.edu.au" class="email">su.s@wehi.edu.au</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Shian Su.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
